src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -67,652 +62,1275 @@ public static unsafe int SequenceCompareTo(ref char first, int firstLength, ref
+                        matches |= Sse2.MoveMask(Sse2.CompareEqual(values2, search).AsByte());
+                        matches |= Sse2.MoveMask(Sse2.CompareEqual(values3, search).AsByte());
+                        if (matches == 0)
+                        {
+                            // Zero flags set so no matches
+                            offset += Vector128<ushort>.Count;
+                            continue;
+                        }
+
+                        // Find bitflag offset of first match and add to current offset, 
+                        // flags are in bytes so divide for chars
+                        return offset + (BitOps.TrailingZeroCount(matches) / sizeof(char));
+                    }
+
+                    if (offset < length)
+                    {
+                        lengthToExamine = length - offset;
+                        goto SequentialScan;
+                    }
+                }
+            }
+            else if (Vector.IsHardwareAccelerated)
+            {
+                if (offset < length)
                 {
-                    // Get the highest multiple of Vector<ushort>.Count that is within the search space.
-                    // That will be how many times we iterate in the loop below.
-                    // This is equivalent to: length = Vector<ushort>.Count * ((int)(pEndCh - pCh) / Vector<ushort>.Count)
-                    length = (int)((pEndCh - pCh) & ~(Vector<ushort>.Count - 1));
+                    lengthToExamine = GetCharVectorSpanLength(offset, length);
 
-                    // Get comparison Vector
                     Vector<ushort> values0 = new Vector<ushort>(value0);
                     Vector<ushort> values1 = new Vector<ushort>(value1);
                     Vector<ushort> values2 = new Vector<ushort>(value2);
                     Vector<ushort> values3 = new Vector<ushort>(value3);
 
-                    while (length > 0)
+                    while (lengthToExamine > offset)
                     {
-                        // Using Unsafe.Read instead of ReadUnaligned since the search space is pinned and pCh is always vector aligned
-                        Debug.Assert(((int)pCh & (Unsafe.SizeOf<Vector<ushort>>() - 1)) == 0);
-                        Vector<ushort> vData = Unsafe.Read<Vector<ushort>>(pCh);
-                        var vMatches = Vector.BitwiseOr(
+                        Vector<ushort> search = LoadVector(ref searchSpace, offset);
+                        var matches = Vector.BitwiseOr(
                                             Vector.BitwiseOr(
-                                                Vector.BitwiseOr(Vector.Equals(vData, values0), Vector.Equals(vData, values1)),
-                                                Vector.Equals(vData, values2)),
-                                            Vector.Equals(vData, values3));
-
-                        if (Vecto
```

Comment:
Does it need a `VZEROUPPER` here? (There is no intrinsic to manually issue one)
```asm
G_M46985_IG09:
       movsxd   rax, edi
       vmovupd  ymm5, ymmword ptr[rcx+2*rax]
       vpcmpeqw ymm6, ymm0, ymm5
       vpmovmskb eax, ymm6
       vpcmpeqw ymm6, ymm1, ymm5
       vpmovmskb r14d, ymm6
       or       eax, r14d
       vpcmpeqw ymm6, ymm2, ymm5
       vpmovmskb ebp, ymm6
       or       eax, ebp
       vpcmpeqw ymm6, ymm3, ymm5
       vpmovmskb ebp, ymm6
       or       eax, ebp
       vpcmpeqw ymm6, ymm4, ymm5
       vpmovmskb ebp, ymm6
       or       eax, ebp
       test     eax, eax
       jne      G_M46985_IG11
       add      edi, 16
       cmp      ebx, edi
       jg       SHORT G_M46985_IG09

G_M46985_IG10:
       mov      eax, esi
       sub      eax, edi
       mov      ebx, eax
       and      ebx, -8
       cmp      ebx, edi         ; if (lengthToExamine > offset)
       jle      G_M46985_IG13
       movzx    r14, dx
       vmovd    xmm0, r14d       ; Vector128<ushort> values0 = Vector128.Create(value0);
       vpbroadcastw xmm0, xmm0
       movzx    rax, r8w
       vmovd    xmm1, eax
       vpbroadcastw xmm1, xmm1
       movzx    rax, r9w
       vmovd    xmm2, eax
       vpbroadcastw xmm2, xmm2
       movzx    rax, r10w
       vmovd    xmm3, eax
       vpbroadcastw xmm3, xmm3
       movzx    rax, r11w
       vmovd    xmm4, eax
       vpbroadcastw xmm4, xmm4
       movsxd   rax, edi
       vmovupd  xmm5, xmmword ptr [rcx+2*rax]
       vpcmpeqw xmm0, xmm0, xmm5
       vpmovmskb eax, xmm0
       vpcmpeqw xmm0, xmm1, xmm5
       vpmovmskb ebx, xmm0
       or       eax, ebx
       vpcmpeqw xmm0, xmm2, xmm5
       vpmovmskb r14d, xmm0
       or       eax, r14d
       vpcmpeqw xmm0, xmm3, xmm5
       vpmovmskb ebx, xmm0
       or       eax, ebx
       vpcmpeqw xmm0, xmm4, xmm5
       vpmovmskb ebx, xmm0
       or       eax, ebx
       test     eax, eax
       jne      SHORT G_M46985_IG12
       add      edi, 8
       jmp      SHORT G_M46985_IG13
```

Reply:
At least I don't think there is a VZEROUPPER api which I could use?

Reply:
Alternative would be to foldback from end for the last bits; so its either `xmm` or `ymm` but not both; however that would make the last one unaligned more often than not (i.e 15 out of 16 times); whereas currently (GC aside) it should be always aligned.

Reply:
Raised issue: "API Proposal: Avx.Zeroupper " https://github.com/dotnet/corefx/issues/34875

Reply:
> Does it need a VZEROUPPER here?

No, that's for mixing non-VEX SSE code with AVX code. That's not possible in .NET, except at method boundaries.

Reply:
Ah, ok

Reply:
Ah, ok

Reply:
Is `xmm6` special in anyway?

Wondering why it adds this at the end
```asm
G_M46985_IG19:
       8BC7                 mov      eax, edi

G_M46985_IG20:
       C5F8283424           vmovaps  xmm6, qword ptr [rsp] ; ?
       C5F877               vzeroupper 
       4883C410             add      rsp, 16
       5B                   pop      rbx
       5D                   pop      rbp
       5E                   pop      rsi
       5F                   pop      rdi
       415E                 pop      r14
       C3                   ret  
```

Reply:
To trigger compilation of the method
```csharp
"Hello World!".IndexOfAny(new[] { 'a', 'b', 'c', 'd', 'e' });
```

Reply:
> Is xmm6 special in anyway?

Call preserved register.

Reply:
Changed the shorter param `.IndexOfAny`s  https://github.com/dotnet/coreclr/pull/22187/commits/848ac58e52a7ee13995e0d9d7d35f28e88cc53f9

e.g. for 3 chars
```asm
G_M46981_IG08:
       movsxd   rax, r11d
       vmovupd  ymm3, ymmword ptr[rcx+2*rax]
       vpcmpeqw ymm4, ymm0, ymm3
       vpcmpeqw ymm5, ymm1, ymm3
       vpcmpeqw ymm3, ymm2, ymm3
       vpor     ymm4, ymm4, ymm5
       vpor     ymm3, ymm4, ymm3
       vpmovmskb eax, ymm3
       test     eax, eax
       jne      SHORT G_M46981_IG10
       add      r11d, 16
       cmp      esi, r11d
       jg       SHORT G_M46981_IG08
```

The longer one 5 (chars); is more commonly used in corefx for validating whether there are escapable chars in xml elements, path search params, etc so will be shorter strings in that regard.

Reply:
`xmm` version being similar:
```asm
vmovupd  xmm3, xmmword ptr [rcx+2*rax]
vpcmpeqw xmm0, xmm0, xmm3
vpcmpeqw xmm1, xmm1, xmm3
vpcmpeqw xmm2, xmm2, xmm3
vpor     xmm0, xmm0, xmm1
vpor     xmm0, xmm0, xmm2
vpmovmskb eax, xmm0
test     eax, eax
jne      SHORT G_M46981_IG11
add      r11d, 8
jmp      SHORT G_M46981_IG12
```

Reply:
`xmm` uses less registers as `ymm` has to loop and keep the values to compare to.

Reply:
Can apply same thing for `xmm` path in 4,5 params https://github.com/dotnet/coreclr/pull/22187/commits/6b9b72323658e3c27ad109a06b3ba072236fa07e which gives
```asm
vmovupd  xmm5, xmmword ptr [rcx+2*rax]
vpcmpeqw xmm0, xmm0, xmm5
vpcmpeqw xmm1, xmm1, xmm5
vpcmpeqw xmm2, xmm2, xmm5
vpcmpeqw xmm3, xmm3, xmm5
vpcmpeqw xmm4, xmm4, xmm5
vpor     xmm0, xmm0, xmm1
vpor     xmm1, xmm2, xmm3
vpor     xmm0, xmm0, xmm4
vpor     xmm0, xmm0, xmm1
vpmovmskb eax, xmm0
test     eax, eax
jne      SHORT G_M46985_IG12
add      edi, 8
jmp      SHORT G_M46985_IG13
```

Reply:
Did you measure perf?

Reply:
IndexOfAny_5 cross over is between 256 and 1023 elements https://github.com/dotnet/coreclr/pull/22187#issuecomment-460068120 and then only slight